### ISSUE TITLE
Campaigns and/or other no-respawns rounds never end with all characters dead and endroundatlevelend = false

### ISSUE THREAD

**Issue by NilanthAnimosus** [Regalis11/Barotrauma#1319]
- [x] I have searched the issue tracker to check if the issue has already been reported.

**Description**
so I wanted endroundatlevelend=false so the level does not end the moment the level end is well, reached. but it created another hilarious problem for my server.

if there are no respawns, such as a campaign level or mission with respawns off, and there are no connected living players left - the round just continues infinitely.

In fact, having the only character in there suffocate (myself) after having disconnected and come back, its still running. it can run for hours this way. new players can only spectate and cannot vote to end an effectively empty server.

returning players also I believe cannot vote to end. this is a situation that bricks the server unless someone has the permission "End Round".

**Version**
0.8.9.6 steam closed alpha

**Additional information**

The behavior I really wanted was for a round that can end when everyone is dead as normal, but not when the end of the level is reached, but still capable of being ended through in-game means such as the watcher and a more natural feeling gameplay loop.

it would be nice if the last players (say a syncing error occurs) got disconnected, even if others were spectating - the round continues in a sort of grace period to rejoin. mind you because of the nature of campaign I had set the killdisconnected timers to very large values alongside kickafk to prevent death of a character simply because one goes AFK or loses sync in a round - even allowing that character to stay alive for a whole 30 minutes to an hour if need be.

could we at least get a behaviour where if the only players who joined are spectating and wont be respawning, -and- there are no players actively controlling a character to end the round? or allow spectators the option to vote to end where if there -are no playing living characters- this counts or is permitted?

the moment the last player leaves, it is officially a bricked server for every player

**Comment by NilanthAnimosus** [comment]
by adding nothing more than a quickly implemented timer into gameserver, checking if its over 60f, and basing it on crewdead (which only checks connected clients) for beyond a minute and making it end the game off that

Pretty much gave me the quick grace period I fancied without leaving an empty / fully dead server stuck on a round without respawns.

```csharp
                if (isCrewDead)
                {
                    if(RestartTimer == 0f)
                    {
                        SendChatMessage("No living clients detected and respawns disabled.", ChatMessageType.Server, null, null);
                        SendChatMessage("If no characters revive / rejoin the server will end round in 60 seconds.", ChatMessageType.Server, null, null);
                    }
                    RestartTimer += deltaTime;
                }
                else
                {
                    RestartTimer = 0f;
                }
```

```
                //restart if all characters are dead or submarine is at the end of the level
                //or when no remaining living connected clients remain with no respawns
                if ((serverSettings.AutoRestart && isCrewDead)
                    ||
                    (serverSettings.EndRoundAtLevelEnd && subAtLevelEnd)
                    || (RestartTimer >= 60f && respawnManager != null))
                {
                    if (serverSettings.AutoRestart && isCrewDead)
                    {
                        Log("Ending round (entire crew dead)", ServerLog.MessageType.ServerMessage);
                    }
                    else if(RestartTimer >= 60f)
                    {
                        Log("Ending round (no living players left connected)", ServerLog.MessageType.ServerMessage);
                    }
                    else
                    {
                        Log("Ending round (submarine reached the end of the level)", ServerLog.MessageType.ServerMessage);
                    }
```

**Comment by Regalis11** [comment]
Fixed in 05cc34a
